src/preload/syscall_hook.S

```diff
@@ -572,6 +572,440 @@ SYSCALLHOOK_START(_syscall_hook_trampoline_49_89_ca)
 	call __morestack
 SYSCALLHOOK_END(_syscall_hook_trampoline_49_89_ca)
 
-#endif /* __x86_64__ */
+
+#elif defined(__aarch64__)
+        .text
+
+        .set syscallbuf_stub_alt_stack, preload_thread_locals
+        .set stub_scratch_1, preload_thread_locals + 16
+        .set alt_stack_nesting_level, preload_thread_locals + 24
+        .set stub_scratch_2, preload_thread_locals + 8 * 13
+
+        .p2align 4
+_syscallbuf_code_start:
+
+_syscall_hook_trampoline:
+        // stack frame:
+        // 208-688: q2 - q31
+        // 128-200: x10 - x18
+        // 112-128: x7, x9
+        // 104-112: x6
+        // 48-104: syscall_info
+        // 32-48: x29, x30
+        // 0-32: q0, q1
+        .cfi_startproc
+        stp     q0, q1, [sp, -688]!
+        .cfi_def_cfa_offset 688
+        .cfi_rel_offset q0, 0
+        .cfi_rel_offset q1, 0 + 16
+        stp     x29, x30, [sp, 32]
+        .cfi_rel_offset x29, 32
+        .cfi_rel_offset x30, 32 + 8
+        /* Build a |struct syscall_info| on the stack by pushing the arguments
+           and syscall number. */
+        stp     x8, x0, [sp, 48]
```

Comment:
Factor out this pattern (with the cfis) into a `#define`, e.g. `STORE_PAIRX(x8, x0, 48)` to avoid any accidental future copy/paste bugs if we ever need to adjust it?

Reply:
You just mean the

```asm
stp xm, xn, [sp, a]
.cfi_rel_offset xm, a
.cfi_rel_offset xn, a + 8
```

pattern right? Yeah, sure. Now that I know the patterns that I used a lot I'll convert some of those into macros. I would maybe also define two different ones for x and q and pass in only the register numbers since otherwise you could accidentally do a `STORE_PAIRX(q0, q1, 20)` and get the wrong offset for q1...